consensus/src/consensus_observer/observer/execution_pool.rs

```diff
@@ -62,3 +69,545 @@ impl ObservedOrderedBlock {
         }
     }
 }
+
+/// Returns all ordered blocks for the given ordered block with window.
+/// This requires traversing backward via the parent links to identify
+/// and fetch the blocks from the pending block store. The blocks are
+/// returned in chronological order, and if any block is missing, this
+/// will return None.
+pub fn get_all_blocks_for_window(
+    pending_block_store: Arc<Mutex<PendingBlockStore>>,
+    ordered_block: &OrderedBlock,
+    window_size: u64,
+) -> Option<Vec<OrderedBlock>> {
+    // If the window size is 0, something is wrong. Log the error and return.
+    if window_size == 0 {
+        error!(
+            LogSchema::new(LogEntry::ConsensusObserver).message(&format!(
+                "Execution pool window size is 0 for ordered block with window: {:?}",
+                ordered_block.proof_block_info()
+            ))
+        );
+        return None;
+    }
+
+    // TODO: this doesn't support ordered blocks with multiple inner blocks.
+    // But, this should be fine, as we'd fallback to state sync anyway.
+
+    // Identify the window boundary (i.e., the lowest round that falls inside the window)
+    let ordered_block_round = ordered_block.first_block().round();
+    let window_boundary_round = ordered_block_round
```

Comment:
There's now a util function
`let window_start_round = calculate_window_start_round(ordered_block_round, window_size);`

Reply:
Done (reused the existing logic!)